### ISSUE TITLE
SvelteKit extending User type does not save the extended values into session

### ISSUE THREAD

**Issue by Luxuride** [nextauthjs/next-auth#6889]
### Question 💬

This question is possibly related to #6882 but I'm using `'@auth/core/types'` instead of `'next-js'` and SvelteKit and not react, so the issue might be fundamentally different. Please close this issue and link it in that ticket if it is fundamentally the same issue. However I think there might be some differences.

I have problem in SvelteKit that even when I extend my User type, and types surrounding it to not get TS error. The added type does not show up in `await event.locals.getSession()` result in `layout.server.ts`.

`hooks.server.ts`
```ts
declare module '@auth/core/types' {
	interface Session {
		user: {
			groups: string[];
		} & DefaultSession['user'];
	}

	interface User {
		groups: string[];
	}

	interface Profile {
		preferred_username: string | undefined;
		groups: string[] | undefined;
	}
}

declare module '@auth/sveltekit' {
	interface SvelteKitAuthConfig {
		providers: Provider<Profile>[];
	}
}

export const handle = SvelteKitAuth({
	providers: [
		{
			id: 'myauth',
			name: 'My Auth',
			type: 'oidc',
			issuer: 'https://myissuer/realms/master',
			wellKnown: AUTH_WELL_KNOWN,
			clientId: AUTH_CLIENT_ID,
			clientSecret: AUTH_CLIENT_SECRET,
			// idToken: true,
			checks: ['pkce', 'state'],
			profile: async (profile, _tokens) => {
				return {
					id: profile.sub ?? '',
					name: profile.preferred_username,
					groups: profile.groups ?? []
				}; //There is only { name: string } in session
			}
		}
	]
});
```
TS tells me everything is type save, but then authjs omits the id and groups.

### How to reproduce ☕️

Extend `User` type, try to return the extended type in `profile` function. Then it will not show up in `await event.locals.getSession()`

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by aakash14goplani** [comment]
Below solution worked for me, check if this is working for you as well.

I am using pnpm as packaging manager.

1. Create a folder called *types* and within that folder add file *auth.d.ts*. Refer project structure below:
      ```
      src
      static
      types
         auth.d.ts
      ```

2. Within *auth.d.ts* file, you can add your types. Below is example for User but you can configure for Profile, Tokens and Accounts as well!
   ```ts
   import type { Your_Custom_Interfaces } from '$src/lib/constants';
   import type { DefaultSession } from '@auth/core/types';

   declare module '@auth/core/types' {
	interface Session {
		user: {} & DefaultSession['user'] & Your_Custom_Interfaces;
	}
   }
   ```

**Comment by Luxuride** [comment]
I tried playing with this solution, but didn't get it working. When I move the `declare module` to `types/auth.d.ts`, the provider in `providers: []` starts saying it does not satisfy the type definitions again and the runtime result is the same. I get only `{ name: string }`

**Comment by aakash14goplani** [comment]
Check solutions in these 2 threads:
https://github.com/nextauthjs/next-auth/issues/6640
https://github.com/nextauthjs/next-auth/issues/6174

**Comment by brendanmatkin** [comment]
I'm having the same issue. I've tried a few of the linked solutions but it still doesn't show up. Using the solution suggested by @aakash14goplani above, I can *cause* ts errors so I suppose ts is recognizing my modified `Session["user"]` type. I was using  

`'@auth/core': ^0.3.0` + `'@auth/sveltekit': ^0.2.0` 
then updated to 
`'@auth/core': ^0.5.1` + `'@auth/sveltekit': ^0.3.0`

..and the issue seems the same for both. I can modify the `DefaultSession` properties (email, name, image) in `session.user`, so I know that I'm modifying it, but can't get it to show anything extra. So long as I add using properties from profile then that property just doesn't show (but the rest work).

Interestingly if I add a property that does not belong to the `OIDCConfig<Auth0Profile>` type (e.g. `return { id: profile.sub, foo: ["bar", "baz"] }`), then the `session.user` object shows up empty. However this might be normal TS behaviour - I'm still getting the hang of it. 

I just had a thought: could this be related to vite using esbuild for typescript compiling instead of tsc?

**Comment by Luxuride** [comment]
I've been playing with trying to send data into the app and I found out you can put all data under { name: any} and { image: any }.
Like
```ts
			profile: async (profile, _token) => {
				return {
					id: profile.sub,
					name: profile,
				};
			}
```
It is an ugly workaround but at least it works.

**Comment by brendanmatkin** [comment]
So we might be doing it wrong - at least I was for what I need (which is to get access to oauth profile data on the session.user). This only works if you are using a JWT session. My guess is that the defaults of these callbacks is what is over-writing the session.user regardless of your User type (for security reasons). The documented way seems to be via [AuthConfig.callbacks](https://authjs.dev/guides/basics/callbacks#jwt-callback). i.e you add the desired properties to the token with `jwt()`, then add values to the session _from the token_ via `session()`.

From the callbacks page: 

> By default, only a subset of the token is returned for increased security. If you want to make something available you added to the token through the jwt() callback, you have to explicitly forward it here to make it available to the client.

So something like this (e.g. to add `session.user.nickname`):
in `src/hooks.server.ts`
```typescript
export const handle = SvelteKitAuth({
  providers: [
    Auth0({
      clientId: PUBLIC_AUTH0_ID,
      clientSecret: AUTH0_SECRET,
      issuer: PUBLIC_AUTH0_DOMAIN,
    }) as Provider,
  ],
  callbacks: {
    jwt({ token, user, profile }) {
      if (user) {
        token.nickname = profile?.nickname;
      }
      return token;
    },
    session({ session, token }) {
      session.user.nickname = token.nickname as string;
      return session;
    },
  },
});
```

created `src/types/auth.d.ts`:
```typescript
export declare module "@auth/core/types" {  // I'm using PNPM but this seems to be working fine
  interface Profile {
    nickname?: string;
  }
  interface Session {
    user: {
      nickname?: string;
    };
  }
}
```

Please let me know if you see anything wrong or suspicious!

**Comment by brendanmatkin** [comment]
> I've been playing with trying to send data into the app and I found out you can put all data under `{ name: any }` and `{ image: any }`. Like
> 
> ```ts
> 			profile: async (profile, _tokens) => {
> 				return {
> 					id: profile.sub,
> 					name: profile,
> 				};
> 			}
> ```
> 
> It is an ugly workaround but at least it works.

@Luxuride good call! This worked for me too until I switched to the solution that I just posted above that I think is more correct. Your idea actually got me hunting for something different that helped me to find it.
